src/seismometer/data/pandas_helpers.py

```diff
@@ -153,11 +175,132 @@ def infer_label(dataframe: pd.DataFrame, label_col: str, time_col: str) -> pd.Da
         dataframe[label_col] = dataframe[label_col].astype(float)
     except BaseException:  # Leave as nonnumeric
         pass
-    dataframe.loc[dataframe[label_col].isna(), label_col] = (
-        dataframe.loc[dataframe[label_col].isna(), time_col].notna().astype(int)
+
+    dataframe.loc[dataframe[label_col].isna(), label_col] = np.where(
```

Comment:
So, using np.select was causing some weird casting issues which turned out to be difficult to get around as you stated.
I can mostly get around the issue if I do something like
```python
label_na_map = dataframe[label_col].isna()
if len(dataframe[label_na_map])==0: #If no missing labels return dataframe as is
      return dataframe
time_na_map = dataframe.loc[label_na_map, time_col].isna()

dataframe.loc[label_na_map, label_col] = np.select(
      [time_na_map],
      [[0] * len(dataframe[label_na_map])], #Where label and time are NA place a 0
      (impute_val or 1) #Where label is NA but time is not, place the impute value or 1 if impute isn't specified
)
dataframe[label_col].replace({"1": 1, "0": 0}, inplace=True)
```
Though that's rather ugly and still requires a replace call.

I've instead opted to just set the values directly into the dataframe instead of using a numpy api. I suspect this might be slightly less performant, but it should still be quick and readable.

https://github.com/epic-open-source/seismometer/blob/4909413838d4a97c0ff81e3de0c5776d1d833642/src/seismometer/data/pandas_helpers.py#L181-L184